### PR TITLE
#19: Follow links

### PR DESCRIPTION
**Pull request by mschmieder**
Hey there,

I took a shot at implementing the `follow links` feature. Again, I'm not a `Golang` expert so please have a good look at the implementation.

I had to refactor the one or the other function a bit, to make it better suitable for the approach I'm using. I decided to do a two-step approach, with

1. Find all relative paths within the markdown files
2. Upload all of those files relative to the file they were mentioned
3. Store the returned urls
4. Patch all markdown files with the retrieved confluence urls
5. Re-upload them to confluence

It is working for me so far, but for sure there might be some corner cases I did not think of. Please test this out and let me know if it works as expected.

Cheers,
Matthias

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,7 +59,8 @@ Usage:
 Flags:
   -d, --debug                Enable debug logging
   -e, --endpoint string      Confluence endpoint. (Alternatively set CONFLUENCE_ENDPOINT environment variable) (default "https://mydomain.atlassian.net/wiki")
-  -x, --exclude strings      list of exclude file patterns (regex) that will be applied on markdown file paths
+  -x, --exclude strings      list of exclude file patterns (regex) for that will be applied on markdown file paths
+      --follow-links         If enabled, documents linked (relative) will be followed and uploaded to conflunece as well
```
Comment (hpierre74) [inline README.md:63]:
```suggestion
      --follow-links         If enabled, documents linked (relative) will be followed and uploaded to confluence as well
```

File: cmd/root.go [hunk 2]
```diff
@@ -23,6 +23,7 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&m.Parent, "parent", "", "Optional parent page to next content under")
 	rootCmd.PersistentFlags().BoolVarP(&m.Debug, "debug", "d", false, "Enable debug logging")
 	rootCmd.PersistentFlags().BoolVarP(&m.UseDocumentTitle, "use-document-title", "", false, "Will use the Markdown document title (# Title) if available")
+	rootCmd.PersistentFlags().BoolVarP(&m.FollowLinks, "follow-links", "", false, "If enabled, documents linked (relative) will be followed and uploaded to conflunece as well")
```
Comment (hpierre74) [inline cmd/root.go:26]:
```suggestion
	rootCmd.PersistentFlags().BoolVarP(&m.FollowLinks, "follow-links", "", false, "If enabled, documents linked (relative) will be followed and uploaded to confluence as well")
```

**Comment by justmiles** [comment]
@mschmieder thanks for contributing, and sorry for this super delayed response. Could you check out some of the comments that @hpierre74 made?

I'll test these changes and we'll get this merged.

**Comment by cprobbie** [comment]
Hi @justmiles , thanks for making this tool! I'd love to use it. Could you please merge this PR because I really need this follow links feature?

**Comment by justmiles** [comment]
Yes! Sorry for the lack of attention. I'll test it today.

**Comment by justmiles** [comment]
Quick update - I tested this and it's throwing the following an error:

```
Error updating content: com.atlassian.confluence.api.service.exceptions.BadRequestException: Content body cannot be converted to new editor format
```

I haven't dug into where this was introduced yet, but trying to get this sorted before merging.

**Comment by sanscontext** [comment]
This would be super helpful if this allows attachments - is there anything I can do to help out here?

**Comment by justmiles** [comment]
This needs more testing. I spent a little bit of time on it and I had negative results, but I haven't dug into why.  I got down a rabbit hole writing the unit test, but I haven't finished.

If you want to help out you can run this branch against your own documents to try and sniff out any issues.

Thanks!

**Comment by alfechner** [comment]
May I kindly ask about the state of this? We're desperately waiting for this feature.

**Comment by justmiles** [comment]
Sorry for the lack of attention. Lots going on and this has been on the back-burner. Definitely looking for help if someone has the time.

**Comment by matsu2254** [comment]
Hey there.
Thanks for the great changes.

I tried this and it worked(confluence cloud).

However, the relative path images are no longer displayed.
This problem is related to the creation of a temporary file.
Passing the original file path seems to solve the problem.
Here is the difference from the "master branch" which also adapted that change.
[diff.patch](https://github.com/justmiles/go-markdown2confluence/files/11714389/diff.patch)

**Comment by justmiles** [comment]
This was a great approach! I was able to resolve this through Goldmark instead of copying the raw markdown file. There's still an issue that it essentially requires two syncs, but following links is supported in [v4.0.0-alpha-2](https://github.com/justmiles/go-markdown2confluence/releases/tag/v4.0.0-alpha-2).
